src/dataflow/src/render/reduce.rs

```diff
@@ -208,144 +248,267 @@ where
             // Distinct is a special case, as there are no aggregates to aggregate.
             // In this case, we use a special implementation that does not rely on
             // collating aggregates.
-            if aggregates.is_empty() {
-                let (oks, errs) = (
-                    ok_input.reduce_abelian::<_, OrdValSpine<_, _, _, _>>("DistinctBy", {
-                        |key, _input, output| {
-                            output.push((key.clone(), 1));
+
+            // First, let's plan out what we are going to do with this reduce
+            let lir = plan_reduce(aggregates.clone(), *monotonic, *expected_group_size);
+            let arrangement = build_reduce(ok_input, lir);
+            let index = (0..group_key.len()).collect::<Vec<_>>();
+            self.set_local_columns(
+                relation_expr,
+                &index[..],
+                (arrangement, err_input.arrange()),
+            );
+        }
+    }
+}
+
+fn plan_reduce(
+    aggregates: Vec<AggregateExpr>,
+    monotonic: bool,
+    expected_group_size: Option<usize>,
+) -> LIrReduceExpr {
+    let lower = move |typ: ReductionType, aggregates_list: Vec<(usize, AggregateExpr)>| {
+        match typ {
+            ReductionType::Accumulable => LIrReduceExpr::Accumulable(AccumulableReduceExpr {
+                aggrs: aggregates_list,
+            }),
+            ReductionType::Hierarchical => {
+                if monotonic {
+                    LIrReduceExpr::Hierarchical(HierarchicalReduceExpr::Monotonic(aggregates_list))
```

Comment:
i don't love how verbose lines like this end up being and welcome any suggestions.